File: UTCS/UTCS/Classes/Model/UTCSLabViewLayout.m
/*
//
//  UTCSLabsCollectionViewLayout.m
//  UTCS
//
//  Created by Comyar Zaheri on 4/25/14.
//  Copyright (c) 2014 UTCS. All rights reserved.
//

#import "UTCSLabViewLayout.h"
#import "UTCSLabView.h"
#import "UTCSLabViewLayoutAttributes.h"

@interface UTCSLabViewLayout ()

@property (nonatomic) NSDictionary  *meta;
@property (nonatomic) NSDictionary  *values;
@property (nonatomic) NSDictionary  *layoutAttributes;
@property (nonatomic) NSArray       *indexPathLayoutAttributes;
@end


@implementation UTCSLabViewLayout

- (instancetype)initWithFilename:(NSString *)filename
{
    if (self = [super init]) {
        NSString *path = [[NSBundle mainBundle]pathForResource:filename ofType:@"plist"];
        NSDictionary *layoutDictionary = [NSDictionary dictionaryWithContentsOfFile:path];
        
        _meta   = layoutDictionary[@"meta"];
        _values = layoutDictionary[@"values"];
        _numberOfLabMachines = [_values count];
        
    }
    return self;
}

- (NSString *)labMachineNameForIndexPath:(NSIndexPath *)indexPath
{
    UTCSLabViewLayoutAttributes *layoutAttribute = [self layoutAttributesForIndexPath:indexPath];
    for (NSString *key in self.layoutAttributes) {
        if ([layoutAttribute isEqual:self.layoutAttributes[key]]) {
            return key;
        }
    }
    return nil;
}

- (void)prepareLayoutForLabView:(UTCSLabView *)labView
{
    CGFloat heightMultiplier = CGRectGetHeight(labView.bounds) / [self.meta[@"maximumGridHeight"]floatValue];
    CGFloat widthMultiplier = CGRectGetWidth(labView.bounds) / [self.meta[@"maximumGridWidth"]floatValue];
    
    CGFloat height = CGRectGetHeight(labView.bounds) / 28.25;
    CGFloat width = height;
    
    NSUInteger index = 0;
    NSMutableDictionary *layoutAttributes = [NSMutableDictionary new];
    NSMutableArray      *indexPathLayoutAttributes = [NSMutableArray new];
    for (NSString *key in self.values) {
        NSDictionary *machineCoordinates = self.values[key];
        
        CGFloat x = [machineCoordinates[@"x"]floatValue];
        CGFloat y = [machineCoordinates[@"y"]floatValue];
        
        UTCSLabViewLayoutAttributes *attributes = [UTCSLabViewLayoutAttributes new];
        attributes.indexPath = [NSIndexPath indexPathForItem:index inSection:0];
        attributes.center = CGPointMake(x * widthMultiplier, y * heightMultiplier);
        attributes.size = CGSizeMake(width, height);
        
        layoutAttributes[key] = attributes;
        [indexPathLayoutAttributes addObject:attributes];
        index++;
    }
    self.layoutAttributes = layoutAttributes;
    self.indexPathLayoutAttributes = indexPathLayoutAttributes;
}

- (UTCSLabViewLayoutAttributes *)layoutAttributesForIndexPath:(NSIndexPath *)indexPath
{
    return self.indexPathLayoutAttributes[indexPath.row];
}

- (UTCSLabViewLayoutAttributes *)layoutAttributesForLabMachineName:(NSString *)labMachineName
{
    return self.layoutAttributes[labMachineName];
}


@end

File: UTCS/UTCS/Classes/UTCSLabViewLayout.swift
//
//  UTCSLabViewLayout.swift
//  UTCS
//
//  Created by Taylor Schmidt on 4/6/16.
//  Copyright © 2016 UTCS. All rights reserved.
//


//  This file might need some work. I'm not sure if the types are
//  all okay or what we want error handling to look like

import Foundation
import UIKit

class UTCSLabViewLayout {
    var meta: NSDictionary // TODO: What are these types
    var values: NSDictionary
    var layoutAttributes: [String: UTCSLabViewLayoutAttributes] = [:]
    var indexPathLayoutAttributes: [UTCSLabViewLayoutAttributes] = []
    var numberOfLabMachines: Int
    
    init?(filename: String) {
        let p = NSBundle.mainBundle().pathForResource(filename, ofType: "plist")
        guard let path = p else {
            // some error code?
            return nil
        }
        
        let d = NSDictionary(contentsOfFile: path)
        guard let layoutDictionary = d else {
            // some error code?
            return nil
        }
        
        self.meta = layoutDictionary["meta"] as! NSDictionary
        self.values = layoutDictionary["values"] as! NSDictionary
        self.numberOfLabMachines = self.values.count
    }
    
    // not sure about this one. Will these types work out?
    func labMachineNameForIndexPath(indexPath: NSIndexPath) -> String? {
        let layoutAttribute = layoutAttributesForIndexPath(indexPath)
        for (key, _) in self.layoutAttributes {
            // Check this. I don't really know if this is correct
            guard let cur = self.layoutAttributes[key] else {
                // some error code?
                return nil
            }
            // This equality check was especially dubious. Is this supposed to check identity?
            if (layoutAttribute as! AnyObject).isEqual(cur) {
                return key
            }
        }
        return nil
    }
    
    func prepareLayoutForLabView(labView: UTCSLabView) {
        let _h = self.meta["maximumGridHeight"]?.floatValue
        let _w = self.meta["maximumGridWidth"]?.floatValue
        guard let maxGridHeight = _h, maxGridWidth = _w else {
            // some error code?
            return
        }
        let heightMultiplier = CGRectGetHeight(labView.bounds) / CGFloat(maxGridHeight)
        let widthMultiplier = CGRectGetWidth(labView.bounds) / CGFloat(maxGridWidth)
        
        let height = CGRectGetHeight(labView.bounds) / 28.25
        let width = height
        var index = 0
        var layoutAttributes: [String: UTCSLabViewLayoutAttributes] = [:]
        var indexPathLayoutAttributes: [UTCSLabViewLayoutAttributes] = []
        
        for (key, machineCoordinate) in self.values {
            let _x = machineCoordinate["x"]!!.floatValue
            let _y = machineCoordinate["y"]!!.floatValue
            guard let x = _x, let y = _y else {
                // some error code? 
                return
            }
            
            let indexPath = NSIndexPath(forItem: index, inSection: 0)
            let center = CGPointMake(CGFloat(x) * widthMultiplier, CGFloat(y) * heightMultiplier)
            let size = CGSizeMake(width, height)
            let attributes = UTCSLabViewLayoutAttributes(size: size, center: center, indexPath: indexPath)
            
            layoutAttributes[key as! String] = attributes
            indexPathLayoutAttributes.append(attributes)
            index += 1
        }
        self.layoutAttributes = layoutAttributes
        self.indexPathLayoutAttributes = indexPathLayoutAttributes
    }
    
    // Should this return type be optional?
    func layoutAttributesForIndexPath(indexPath: NSIndexPath) -> UTCSLabViewLayoutAttributes? {
        return self.indexPathLayoutAttributes[indexPath.row] 
    }
    
    // Should this return type be optional?
    func layoutAttributesForLabMachineName(labMachineName: String) -> UTCSLabViewLayoutAttributes? {
        return self.layoutAttributes[labMachineName]
    }
}
